[Cocoa] startTime of dynamically added EXT-X-DATERANGE metadata cues is always 0

eric-carlson opened the pull request:
#### b63b9425c88b7c4bc0af59caf9b3241d002842a9
<pre>
[Cocoa] startTime of dynamically added EXT-X-DATERANGE metadata cues is always 0
<a href="https://bugs.webkit.org/show_bug.cgi?id=251541">https://bugs.webkit.org/show_bug.cgi?id=251541</a>
rdar://104142471

Reviewed by Jer Noble.

The start time of a metadata group was calculated incorrectly.

* LayoutTests/http/tests/media/hls/track-in-band-ext-x-date-range-start-time-expected.txt: Added.
* LayoutTests/http/tests/media/hls/track-in-band-ext-x-date-range-start-time.html: Added.
* LayoutTests/http/tests/media/resources/hls/dynamic-ext-date-range.py: Added.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::metadataGroupDidArrive): Correct `startTime`
calculation.

Canonical link: <a href="https://commits.webkit.org/259924@main">https://commits.webkit.org/259924@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bce4212d1c405eec43cb7c490b83fcca47b83b05

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106115 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15168 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38943 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115299 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175376 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110020 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16658 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6396 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98324 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114996 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111871 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12674 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95600 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40173 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94547 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27243 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81852 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8419 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28595 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8913 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5191 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14532 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48139 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6858 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10458 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on LayoutTests/http/tests/media/hls/track-in-band-ext-x-date-range-start-time.html:
```diff
@@ -0,0 +1,44 @@
+                for (let i = 0; i < track.cues.length; i++) {
+                    cue = track.cues[i];
+                    testExpected(`track.cues[${i}].startTime`, 0, (i ? '!=' : '=='));
+                }
```
Is `cue = track.cues[i];` necessary here? Could this be changed to:

```
                testExpected(`track.cues[0].startTime`, 0,));
                for (let i = 1; i < track.cues.length; i++)
                    testExpected(`track.cues[${i}].startTime`, 0, '!='));
```

eric-carlson replied:
Fixed both, thanks.

jernoble reviewed (approved):
r=me with nit

eric-carlson committed:
[Cocoa] startTime of dynamically added EXT-X-DATERANGE metadata cues is always 0
https://bugs.webkit.org/show_bug.cgi?id=251541
rdar://104142471

Reviewed by Jer Noble.

The start time of a metadata group was calculated incorrectly.

* LayoutTests/http/tests/media/hls/track-in-band-ext-x-date-range-start-time-expected.txt: Added.
* LayoutTests/http/tests/media/hls/track-in-band-ext-x-date-range-start-time.html: Added.
* LayoutTests/http/tests/media/resources/hls/dynamic-ext-date-range.py: Added.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::metadataGroupDidArrive): Correct `startTime`
calculation.

Canonical link: https://commits.webkit.org/259924@main
